worker/agent.go

```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
 	return
 }
 
-func (a *agent) Connect() (err error) {
-	a.Lock()
-	defer a.Unlock()
-	a.conn, err = net.Dial(a.net, a.addr)
-	if err != nil {
-		return
-	}
-	a.rw = bufio.NewReadWriter(bufio.NewReader(a.conn),
-		bufio.NewWriter(a.conn))
-	go a.work()
-	return
-}
-
 func (a *agent) work() {
+	a.worker.Log(Info, "Starting agent Work For:", a.addr)
 	defer func() {
 		if err := recover(); err != nil {
-			a.worker.err(err.(error))
+			a.reconnect_error(err.(error))
 		}
 	}()
 
 	var inpack *inPack
 	var l int
 	var err error
 	var data, leftdata []byte
-	for {
-		if !a.worker.isShuttingDown() {
-			if data, err = a.read(); err != nil {
-				if opErr, ok := err.(*net.OpError); ok {
-					if opErr.Temporary() {
-						continue
-					} else {
-						a.disconnect_error(err)
-						// else - we're probably dc'ing due to a Close()
-
-						break
-					}
-
-				} else if err == io.EOF {
-					a.disconnect_error(err)
-					break
-				}
-				a.worker.err(err)
-				// If it is unexpected error and the connection wasn't
-				// closed by Gearmand, the agent should close the conection
-				// and reconnect to job server.
-				a.Close()
-				a.conn, err = net.Dial(a.net, a.addr)
-				if err != nil {
-					a.worker.err(err)
+	startConnVersion := a.connectionVersion
+
+	// exit the loop if connection has been replaced because reconnect will launch a new work() thread
+	for startConnVersion == a.connectionVersion && !a.worker.isShuttingDown() {
+
+		if data, err = a.read(); err != nil {
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Temporary() {
+					a.worker.Log(Info, "opErr.Temporary():", a.addr)
+					continue
+				} else {
+					a.worker.Log(Info, "got permanent network error with server:", a.addr, "comm thread exiting.")
+					a.reconnect_error(err)
+					// else - we're probably dc'ing due to a Close()
 					break
 				}
-				a.rw = bufio.NewReadWriter(bufio.NewReader(a.conn),
-					bufio.NewWriter(a.conn))
-			}
-			if len(leftdata) > 0 { // some data left for processing
-				data = append(leftdata, data...)
+			} else {
+				a.worker.Log(Info, "got error", err.Error(), "with server:", a.addr, "comm thread exiting...")
+				a.reconnect_error(err)
+				break
 			}
-			if len(data) < rt.MinPacketLength { // not enough data
-				leftdata = data
-				continue
-			}
-			for {
-				if inpack, l, err = decodeInPack(data); err != nil {
-					a.worker.err(err)
-					leftdata = data
+		}
+		if len(leftdata) > 0 { // some data left for processing
+			data = append(leftdata, data...)
+		}
+		if len(data) < rt.MinPacketLength { // not enough data
```

Comment:
You could remove this whole routine if you want to make a significant improvement to this system. :-) It's really silly and almost the same as the parsing in the client.